### PR TITLE
New helper methods

### PR DESCRIPTION
**Pull request by dawnerd**
Adds:
- separate methods for binding actions and elements
- findOne
- findByName
- getOption
- tests
- data-module-global-_/data-action-global-_
- refs to instance stored on Domodule
- getInstance
- discover

### CODE REVIEW

File: lib/domodule.js [hunk 1]
```diff
@@ -1,17 +1,47 @@
-class Domodule {
-  constructor(el) {
+/* eslint no-new:0 */
+
+class Domodule { // eslint-disable-line no-unused-vars
+  constructor(el, requiredOptions = []) {
     this.el = el;
     this.els = {};
     this.options = this.serializeAttrs('module', this.el);
+    this.requiredOptions = requiredOptions;
```
Comment (jgallen23) [inline lib/domodule.js:8]: do we have to pass this in? or can we read from the class definition?

File: lib/domodule.js [hunk 2]
```diff
@@ -1,17 +1,47 @@
...
+    return this;
+  }
+
+  verifyOptions() {
+    if (!Array.isArray(this.requiredOptions)) {
+      return this;
+    }
+
+    for (const requiredOption of Array.values(this.requiredOptions)) {
+      if (!this.options.hasOwnProperty(requiredOption)) {
```
Comment (jgallen23) [inline lib/domodule.js:24]: will this error if the value is blank?

File: lib/domodule.js [hunk 3]
```diff
@@ -46,4 +109,47 @@ class Domodule {
...
+    const found = this.find(selector);
+
+    if (found.length) {
+      return found[0];
+    }
+
+    return null;
+  }
+
+  findByName(name) {
```
Comment (jgallen23) [inline lib/domodule.js:123]: what is the use case here?

File: lib/domodule.js [hunk 4]
```diff
@@ -46,4 +109,47 @@ class Domodule {
...
+    return this.els[name];
+  }
+
+  getOption(option) {
+    return this.options[option];
+  }
+
+  // static methods can't access `this` so they go last
+  static getInstance(name) {
```
Comment (jgallen23) [inline lib/domodule.js:132]: we should add support for also passing in an element node and returning the instance

File: lib/domodule.js [hunk 5]
```diff
@@ -46,4 +109,47 @@ class Domodule {
...
+    return Domodule.refs[name];
+  }
+
+  static discover(el = 'body') {
+    const els = document.querySelectorAll(el);
+
+    for (const matched of els) {
+      const foundModules = matched.querySelectorAll('[data-module]');
+
+      for (const module of foundModules) {
```
Comment (jgallen23) [inline lib/domodule.js]: what about calling it `moduleEl`?  was a little confused when I was reading the code below

**Comment by jgallen23** [comment]
can we ignore `example/dist/test.js`?

File: lib/domodule.js (hunk 1 above)
Comment (dawnerd) [inline lib/domodule.js:8]: We could read it as part of the attributes

File: lib/domodule.js (hunk 2 above)
Comment (dawnerd) [inline lib/domodule.js:24]: it defaults to [] so it shouldn't error. I'll add a test case for it though.

File: lib/domodule.js (hunk 3 above)
Comment (dawnerd) [inline lib/domodule.js:123]: Cleaner testing I guess? Feels weird pulling a property directly from the class.

Comment (jgallen23) [inline lib/domodule.js:123]: You mean something like this.els.title?

Comment (dawnerd) [inline lib/domodule.js:123]: In my test case it's `Domodule.getInstance('ExampleModule').findByName('tester')`

**Pull request merged**